crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -172,17 +199,37 @@ impl<'src> Parser<'src> {
             }
         }
 
-        if !self.eat(TokenKind::Newline) && self.at_compound_stmt() {
-            self.add_error(
-                ParseErrorType::SimpleStmtAndCompoundStmtInSameLine,
-                self.node_range(start),
-            );
+        // Ideally, we should use `expect` here but we use `eat` for better error message. Later,
+        // if the parser isn't at the start of a compound statement, we'd `expect` a newline.
+        if !self.eat(TokenKind::Newline) {
+            if self.at_compound_stmt() {
+                // test_err simple_and_compound_stmt_on_same_line_in_block
+                // if True: pass if False: pass
+                // if True: pass; if False: pass
+                self.add_error(
+                    ParseErrorType::SimpleAndCompoundStatementOnSameLine,
+                    self.current_token_range(),
+                );
+            } else {
+                // test_err multiple_clauses_on_same_line
+                // if True: pass elif False: pass else: pass
+                // if True: pass; elif False: pass; else: pass
+                // for x in iter: break else: pass
+                // for x in iter: break; else: pass
+                // try: pass except exc: pass else: pass finally: pass
+                // try: pass; except exc: pass; else: pass; finally: pass
+                self.expect(TokenKind::Newline);
```

Comment:
If there's no newline and it's not a compound statement, then we'll fallback to using `expect`. The bug which I was talking about on Discord.

Reply:
Nit: Could we push the error directly instead of calling `expect` or does that require a lot of code. The call looks redudant on the first glance (but it's not).

Reply:
I see. Yeah, it does look redundant on first glance. I'll make the change.